anndata/utils.py

```diff
@@ -311,6 +313,14 @@ def convert_dictionary_to_structured_array(source: Mapping[str, Sequence[Any]]):
     return arr
 
 
+def warn_once(msg: str, category: type[Warning], stacklevel: int = 1):
+    warnings.warn(msg, category, stacklevel=stacklevel)
+    if "PYTEST_CURRENT_TEST" not in os.environ:
```

Comment:
Why is this bit necessary? It seems like this is throwing many errors during testing anyways.

Reply:
Note the “not”. It silences itself in non-testing code

Reply:
That's not quite what I was saying. If you take a look in my [comment here](https://github.com/scverse/anndata/pull/1182#pullrequestreview-1667382109) I show that this warning still shows up in tests on main even though an `ignore` [should be present](https://github.com/scverse/anndata/blob/26e909970c7fbb382e8addfaed6ce76cff4a4e74/anndata/_core/aligned_mapping.py#L72C9-L76).

So, why do we need test specific behavior to NOT filter during testing if it seems that this warning was totally fine showing up without it?

I also really don't like the idea of code branches that are specific to whether the tests are running.

Reply:
That ignore is still active in the same test case after the first hit.

> I also really don't like the idea of code branches that are specific to whether the tests are running.

Me neither! Just as I don’t like silently changing global state like we do with that ignore filter. But since we do the latter, we also need to do the former.

Both hacks are just affecting a warning filter, so it’s not that bad, but I’d like to see us removing both (at the same time). The only way I see this happening is by removing the need for “warn once” hack

Reply:
> Just as I don’t like silently changing global state like we do with that ignore filter

It's very targeted to our specific warning, and I believe this is fairly common practice (I believe I figured this out from an issue on an upstream library). It's not great but I think it's better than this.

I'm not up for this change.

Reply:
> It's not great but I think it's better than this.

I’m confused. I added the `if "PYTEST_CURRENT_TEST" not in os.environ` because it’s the only way to tame the `warnings.simplefilter('ignore', ...)` that I can think of. If you have an alternative idea how we can reliably test where that warning is thrown, I’d be happy to switch to it.

I don’t understand your objection. As said:

1. Global changes are bad. But this is just a warning filter, so I see why it’s fine.
2. Changing behavior during tests is bad. But this is the exact same warning filter from 1., so the exact same reasoning applies.

Why differentiate between the two? In both cases it’s “this would be a code smell, if it wasn’t just about warning visibility, which really isn’t a big deal”

Reply:
OK fixed line we talked about: Globally ignore ExperimentalFeatureWarning